src/v2/services/documents.js

```diff
@@ -11,26 +11,63 @@ const documentRoleRepo = require('../connectors/repository/document-roles');
 const rolesRepo = require('../connectors/repository/roles');
 const { mapValidationErrorDetails } = require('../lib/map-error-response');
 const errors = require('../lib/errors');
+const repo = require('../connectors/repository');
+const Boom = require('@hapi/boom');
+const mappers = require('./lib/mappers');
+
+const datesOverlap = (list, item) => {
+  return list.map(row => {
+    const rangeA = moment.range(item.startDate, item.endDate);
+    const rangeB = moment.range(row.startDate, row.endDate);
+    return !!(rangeA.intersect(rangeB));
+  }).includes(true);
+};
 
-/**
- * Create a moment date range
- *
- * @param {String} startDate A start date in the format YYYY-MM-DD
- * @param {String|null} endDate A end date in the format YYYY-MM-DD or null
- */
-const getDateRange = (startDate, endDate) => {
-  const start = moment(startDate, 'YYYY-MM-DD');
-  const end = endDate === null ? null : moment(endDate, 'YYYY-MM-DD');
-  return moment.range(start, end);
+const createDocument = async document => {
+  const { error, value: validatedDocument } = documentValidator.validateDocument(document);
+
+  if (error) {
+    const details = error.details.map(detail => detail.message);
+    throw new errors.EntityValidationError('Document not valid', details);
+  }
+
+  // get the existing documents for the document ref
+  const currentDocuments = await repo.documents.findByDocumentRef(document.regime, document.documentType, document.documentRef);
+  if (datesOverlap(currentDocuments, document)) {
+    throw new errors.ConflictingDataError('Overlapping start and end date for document');
+  }
+  return repo.documents.create(validatedDocument);
```

Comment:
I think here is where you would need to catch known Postgres errors (e.g. constraint violations) and re-throw as service errors.

Reply:
done - thanks for your help